bench/tpcc: refactor tpcc benchmark

mgartner opened the pull request:
#### bench/tpcc: add wait timeouts

The parent/child process synchronization is now more robust with
timeouts for some of the waits.

Release note: None

#### bench/tpcc: make client verify TPCC database exists

Release note: None

#### bench/tpcc: remove unused setupStmtOption type

Release note: None

#### bench/tpcc: simplify server instantiation

Release note: None

#### bench/tpcc: remove unnecessary benchmark closers

Release note: None

#### bench/tpcc: remove unused serverArgs and setupServerOption types

Release note: None

#### bench/tpcc: simplify workload flags

Release note: None

#### bench/tpcc: clean up contexts

Release note: None

#### bench/tpcc: use testfixtures.ReuseOrGenerate for store directory

We now use `testfixtures.ReuseOrGenerate` as a mechanism for reusing a
store directory between benchmark runs. The TPC-C data is automatically
generated if the store directory does not exist, and an existing store
directory is automatically reused if it exists. A user no longer needs
to perform a 2-step process and provide `--generate-store-dir` and
`--store-dir` flags. The name of the store directory includes
`storage.MinimumSupportedFormatVersion` so that TPC-C data will be
regenerated if the storage format changes.

Release note: None

#### bench/tpcc: consolidate utilities in the same file

Release note: None

#### bench/tpcc: rename files

Release note: None

#### bench/tpcc: remove non-test file

Release note: None

#### bench/tpcc: enable synchronization of the raft log

This makes the benchmark more representative of a real-world workload.
It also didn't have make a meaningful improvement for the generation of
TPC-C schema and data, so it's been removed from
`TestInternalGenerateStoreDir` too.

Release note: None

#### bench/tpcc: clean up benchmark names

Release note: None

#### bench/tpcc: simplify closers

Release note: None

mgartner committed:
bench/tpcc: add wait timeouts

The parent/child process synchronization is now more robust with
timeouts for some of the waits.

Release note: None

mgartner committed:
bench/tpcc: make client verify TPCC database exists

Release note: None

mgartner committed:
bench/tpcc: remove unused setupStmtOption type

Release note: None

mgartner committed:
bench/tpcc: simplify server instantiation

Release note: None

mgartner committed:
bench/tpcc: remove unnecessary benchmark closers

Release note: None

mgartner committed:
bench/tpcc: remove unused serverArgs and setupServerOption types

Release note: None

mgartner committed:
bench/tpcc: simplify workload flags

Release note: None

mgartner committed:
bench/tpcc: clean up contexts

Release note: None

mgartner committed:
bench/tpcc: use testfixtures.ReuseOrGenerate for store directory

We now use `testfixtures.ReuseOrGenerate` as a mechanism for reusing a
store directory between benchmark runs. The TPC-C data is automatically
generated if the store directory does not exist, and an existing store
directory is automatically reused if it exists. A user no longer needs
to perform a 2-step process and provide `--generate-store-dir` and
`--store-dir` flags. The name of the store directory includes
`storage.MinimumSupportedFormatVersion` so that TPC-C data will be
regenerated if the storage format changes.

Release note: None

mgartner committed:
bench/tpcc: consolidate utilities in the same file

Release note: None

mgartner committed:
bench/tpcc: rename files

Release note: None

mgartner committed:
bench/tpcc: remove non-test file

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147125)
<!-- Reviewable:end -->

mgartner commented:
A lot of this is mechanical changes. It's best reviewed one commit at a time.

I think @cockroach-dev-inf only needs to look at the commit: [cmd/dev: add --rewrite flag for bench](https://github.com/cockroachdb/cockroach/pull/147125/commits/e60f806264602c92a5d92d7968dbbb3657d60354), possibly [bench/tpcc: move store directory to testdata](https://github.com/cockroachdb/cockroach/pull/147125/commits/e2062c8a74344d6c59b1f59cb0a26497dd11ec24) to understand the motivation for `--rewrite` for benchmarks. The TLDR is that `--rewrite` allows us to write and reuse benchmark assets in the benchmark's `testdata` directory. It is used to avoid the expensive setup cost for these TPCC benchmarks—rather than rebuild the schema and populate tables we can run a CRDB server with a copy of store files that were previously generated. This seemed like a good option because it is consistent with tests and it should work better with Bazel's sandboxing than allowing the user to pick any arbitrary directory to write these assets to, as was allowed before.

tbg commented on pkg/bench/tpcc/tpcc_bench_test.go:
```diff
@@ -25,42 +20,55 @@ import (
-	defer cleanup()
+	// Generate TPCC data if requested.
+	if *rewrite {
+		storeDir := datapathutils.TestDataPath(b, storePath)
```
a bit of a random place to leave this comment, but currently (I think) this does single-warehouse. Is that sensible? Wouldn't there be a ton of contention? In particular, I think newOrder requires a sequential ID allocation per warehouse, which would become a bottleneck. I'm not sure this is primarily what we want to measure?

(Just want to start the discussion, not necessarily have it here).

mgartner replied:
The client queries are run in a single thread, so there should be no concurrent queries. And that's by design for this micro-benchmark. I want it to show the cost of a query in a more-or-less ideal case, e.g., without contention.

tbg commented on pkg/bench/tpcc/tpcc_bench_test.go:
```diff
@@ -25,42 +20,55 @@ import (
+		if err := os.RemoveAll(storeDir); err != nil {
+			b.Fatalf("failed to clear store dir: %s", err)
+		}
+		cmd, stdout := generateStoreDir.exec()
```
Is there any reason why we're shelling out here? It seems a little circuitous. We spawn a subprocess, which in turn runs CRDB and spawns another subprocess running a client command to populate data?

tbg replied:
You're right :thumbsup: 
<!-- Reviewable comment -OQx-Seh-4qzBmyjFa0R:bma0j5g -->
<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
Thank you for the thorough cleanups and simplifications, and for making them so pleasant to review!

tbg commented:
Some notes playing with this. Entry point was

```
$ ./dev bench ./pkg/bench/tpcc --filter BenchmarkTPCC/literal/newOrder=1 --test-args '-test.cpu=""'
--- FAIL: BenchmarkTPCC/literal/mix=newOrder=1
    bench_test.go:107: waiting on client timed-out:
        === RUN   TestInternalRunClient
        Max Total connections 1 Max connections per pool 20 
        Initializing 1 connections...
        Initializing 0 idle connections...
        Initializing 1 workers and preparing statements...
            internal_test.go:88: tpcc database does not exist, try running with --rewrite first
        --- FAIL: TestInternalRunClient (0.01s)
        FAIL
```

This strikes me as... not ideal. I understand how we got here: with `dev` we can't make a reusable artifact "on demand". But I still think it would be better if the test didn't fail if the fixtures aren't there. @herkolategan might have opinions on what's best here - ideally the vanilla invocation "just works" so that the nightly bench suite doesn't get tripped up here. OTOH it's also useful not to repeatedly set up the fixtures, and to keep them stable between iterations.

Also it's annoying that `newOrder=1` is also a prefix of `newOrder=10,...` (another datapoint that we should maybe just name the default mix `defaultMix` instead of spelling it out). The long mix name will also absolutely mess up any benchstat invocation it is a part of.

I think you need to bump the DEV_VERSION to pick up the `-rewrite` flag.

```
$ ./dev bench ./pkg/bench/tpcc --rewrite --filter BenchmarkTPCC --test-args '-test.cpu=""'
ERROR: unknown flag: --rewrite
$ ./dev build dev
# now it works
```

tbg reviewed (changes requested):
ended up having some more substantial questions, and it's worth at least making sure that the benchmark as is doesn't error out during nightlies.

mgartner committed:
bench/tpcc: enable synchronization of the raft log

This makes the benchmark more representative of a real-world workload.
It also didn't have make a meaningful improvement for the generation of
TPC-C schema and data, so it's been removed from
`TestInternalGenerateStoreDir` too.

Release note: None

mgartner committed:
bench/tpcc: clean up benchmark names

Release note: None

mgartner commented on pkg/bench/tpcc/tpcc_bench_test.go:
```diff
@@ -121,13 +124,15 @@ func (bm *benchmark) run(b *testing.B) {
 	// Reset the timer when the client starts running queries.
-	s.wait()
+	if timedOut := s.waitWithTimeout(); timedOut {
+		b.Fatalf("waiting on client timed-out:\n%s", stdout.String())
```
Done.

mgartner commented on pkg/bench/tpcc/subprocess_commands_test.go:
```diff
@@ -74,6 +75,17 @@ func TestInternalRunClient(t *testing.T) {
+
+	// Verify the TPC-C database exists.
+	if _, err := conn.Exec(context.Background(), "USE "+databaseName); err != nil {
+		t.Fatal(databaseName + " database does not exist")
```
If, for some reason, the store directory was created but the database is empty (i.e., the TPC-C schema was not created), then invoking the `WorkerFns[0]` will hang forever as it continually tries and fails to open a connection because preparing the TPCC queries (which happens when a connection is created) will all fail. I couldn't find an obvious way to get feedback from the workload generator about this type of failure mode, so this quick check for the tpcc database seemed like a reasonable and quick gut-check that database is setup correctly.

tbg replied:
:thumbsup: 
<!-- Reviewable comment -OQx-yZN7N87PHwJF3Bj:b4c9lyr -->
<!-- Sent from Reviewable.io -->

mgartner commented on pkg/bench/tpcc/tpcc_bench_test.go:
Ya I don't think we want this at all during the benchmark. Removed.

mgartner commented:
> Also it's annoying that newOrder=1 is also a prefix of newOrder=10,...

Fixed in a newly added commit.

---

> I think you need to bump the DEV_VERSION to pick up the -rewrite flag.

TIL, thank you! Done.

---

> BenchmarkSysbench grabs really clean allocs and mutex profiles (that don't contain the server setup etc) [here](https://github.com/cockroachdb/cockroach/blob/b710dd4c3803e9f2f43a8a17421fd2674c57a384/pkg/sql/tests/sysbench_test.go#L901). We may want to also adopt this here.

> A bigger thing I missed during review (since it wasn't a decision you made) is that the test runs against a single server. Single servers are quite different from clusters (raft replication traffic etc), so it would be good to get closer to BenchmarkSysbench in that there should at least be a 3node flavor. We're not going to get the "KV/" flavors with the current approach, but that's fine. Covering the SQL/{1node_{local,remote},3node} flavors would be good.

Yes! I'm planning on doing both but wanted to clean up this package before really getting cooking. IMO I don't see a strong need for the `1node` variants. I'd prefer the simplicity of a single, `3node` variant, unless someone feels strongly that that's not sufficient.

---
  
> BenchmarkSysbench also runs a few warmup iterations. This would not be at the top of my list to do here, but it may be worth it at some point

Ya, I'd like to only add it if we think we really need it, in order to keep the benchmark as quick as possible for a faster feedback loop.

---

> I didn't see anything about workload concurrency...

> I would also display the warehouse count in the test configuration...

I'd prefer to optimize these benchmarks for fast feedback loops, so you can quickly check if your small change makes a measurable impact. I think that's the main value of these benchmarks in contrast to the slower, noisier full TPC-C workload/roachtest. I don't think the micro-benchmarks need to perfectly mimic adverse conditions like contention in order for them to help us make perf gains. I personally haven't found that the parallel sysbench benchmarks add much value over the non-parallel ones. And I question whether we'll actually ever need more than one warehouse here. I'm happy to be convinced otherwise though—maybe this is a very SQL-centric perspective?

---

> @mgartner I found this:

Thanks for the pointer. I'll play around with this some more. You bring up some important UX sharp edges.

mgartner commented:
`testfixtures.ReuseOrGenerate` solves most of the previous issues. The benchmark has been updated so that it automatically generates and reuses the store directory. I've backed out the changes to `dev bench` since they are no longer needed.

> Since these fixtures are store dirs, we can't reuse them across SHAs, but we could put the build SHA in there (if linked in) and otherwise use a unique name (preventing reuse across process boundaries).

Unfortunately, the git SHA is not easily obtained—it's only included in `build.Info` for release builds. I've added `storage.MinimumSupportedFormatVersion` to the store directory name, but that's not perfect.

A flag to force regeneration of the storage directory would probably be a good thing to have, too.

tbg reviewed (approved):
Some more comments (and I may not have reviewed every commit, due to the force push and not using reviewable during my first review pass, so if there's anything particular you'd like me to look closely at let me know). Nothing to block the merge!

> Reviewed 2 of 9 files at r19, 2 of 5 files at r20, 1 of 4 files at r23, 4 of 4 files at r24, 6 of 6 files at r25, 4 of 4 files at r26, 7 of 7 files at r27, 2 of 2 files at r28, 4 of 4 files at r29, 3 of 3 files at r30, 2 of 2 files at r31, 2 of 2 files at r32, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147125#-:-OQx2puV8M5GWn9dk4Ql:b-cgzsjt)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @ajstorm, @golgeek, and @herkolategan)

___
*[`pkg/bench/tpcc/bench_test.go` line 16 at r32](https://reviewable.io/reviews/cockroachdb/cockroach/147125#-OQx2Qhm4nptA4yt2HLm:-OQx2Qhm4nptA4yt2HLn:b-prmcxm) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed0b42dc306e90ac296687c0cd4200e81a27ffe1/pkg/bench/tpcc/bench_test.go#L16)):*
Another comment to highlight this bit from my message, which I think you may not have noticed/not responded to:

> The more I play around with this, the more confused I am also about the value of spawning these subprocesses. You've already stripped out some of the complexity, but it's still an all around strange test design, since at no point (thank dog) do we invoke the workload binary - we call into it as a library. Other than riffing off of what was already there, why do we prefer this design over "simply" running the client directly? I think we could dispense with this entire synchronizer business, and wouldn't end up running the test in a `*testing.T` function but would have the proper `testing.B` around which is better since we'll also want the counterpart of `BenchmarkSysbenchParallel`.  
It might be mildly useful to have the profile only reflect the server and not the SQL client, but I'm not at all convinced that this is worth the additional complexity - the client has never played a role in the profiles for sysbench, and I don't expect that to change; and should it change, we'd really want to know about it since these processes do share the underlying machine.

It's not something I'd block this PR on, but I want to get your take on it at least. This test started out really over-engineered, and I think this subprocess thing is part of that. Future generations may thank us (you) for removing it, really. If your take is "meh, I agree, but I don't want to spend more time code golfing" - that's fine by me.
___
*[`pkg/bench/tpcc/tpcc_bench_test.go` line 15 at r19](https://reviewable.io/reviews/cockroachdb/cockroach/147125#-OQwxr144k6iq3fceuFQ:-OQwxr144k6iq3fceuFR:b7x7zcn) ([raw file](https://github.com/cockroachdb/cockroach/blob/7133c6842ef08e9440a5641fcbe4df5e6ee70c11/pkg/bench/tpcc/tpcc_bench_test.go#L15)):*
Adding a comment thread here so we can have a better back and forth on this:

> I'd prefer to optimize these benchmarks for fast feedback loops, so you can quickly check if your small change makes a measurable impact. I think that's the main value of these benchmarks in contrast to the slower, noisier full TPC-C workload/roachtest. I don't think the micro-benchmarks need to perfectly mimic adverse conditions like contention in order for them to help us make perf gains. I personally haven't found that the parallel sysbench benchmarks add much value over the non-parallel ones. And I question whether we'll actually ever need more than one warehouse here. I'm happy to be convinced otherwise though—maybe this is a very SQL-centric perspective?

+1 on fast feedback loops. I wasn't suggesting that this should become a heavyweight test. I find `BenchmarkParallelSysbench` quite useful, but I have also spent time on mutex contention, for which a single-threaded benchmark obviously makes no sense. Additionally, a concurrent benchmark can do a better job at highlighting the impact of garbage collection (since garbage is created at higher rates, and GC runs more frequently).

We didn't add `BenchmarkParallelSysbench` at first either, though, so no pressure to do so now - but it would be good if there weren't obvious issues with doing so later that couldn't easily be solved now. I'm not aware of any.

> I don't think the micro-benchmarks need to perfectly mimic adverse conditions like contention in order for them to help us make perf gains

I was actually worried about the opposite. Concurrency is admittedly a bit more heaviweight, but it also makes the benchmark more realistic. I was worried that if we added concurrency without also incrementing the warehouse count, we'd basically test the performance of incrementing a single key (since I think the order IDs are pulled from a per-warehouse counter). But since we're not adding concurrency now, we are fine with one warehouse.
And similar to sysbench, if we do add it, it can be under a new top-level test, and that one can be configured with more warehouses.
___
*[`pkg/bench/tpcc/subprocess_commands_test.go` line 98 at r27](https://reviewable.io/reviews/cockroachdb/cockroach/147125#-OQx0t1J79xPXcGf9Nxd:-OQx0t1J79xPXcGf9Nxe:b541me) ([raw file](https://github.com/cockroachdb/cockroach/blob/0afcb3ed2d4b4ee5361fe42e56b9a670419d0844/pkg/bench/tpcc/subprocess_commands_test.go#L98)):*
> ```Go
> 	// Verify the TPC-C database exists.
> 	if _, err := conn.Exec(ctx, "USE "+databaseName); err != nil {
> 		t.Fatal(databaseName + " database does not exist, try running with --rewrite first")
> ```

I thought --rewrite was no longer necessary?
___
*[`pkg/bench/tpcc/bench_test.go` line 96 at r32](https://reviewable.io/reviews/cockroachdb/cockroach/147125#-OQx2-IDAXhOO0JpS1am:-OQx2-IDAXhOO0JpS1an:b-hcdv0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/4c879aa7a1a0d3db7823dad2efc9780d5a0c226a/pkg/bench/tpcc/bench_test.go#L96)):*
> ```Go
> 
> 	// Clone the store dir.
> 	td, engCleanup := testutils.TempDir(b)
> ```

FYI there's also `b.TempDir` which gets auto-removed. (I think this didn't exist when `testutils.TempDir` was introduced).


<!-- Sent from Reviewable.io -->

mgartner committed:
bench/tpcc: simplify closers

Release note: None

mgartner commented on pkg/bench/tpcc/tpcc_bench_test.go:
Done.

mgartner commented:
I think I've addressed all the minor feedback, and this takes us in the right general direction, so I'm going to merge this when I get a passing build. There's still more work to be done though! I'll have a follow-up PR(s) for the things discussed above:

* Removing the subprocess strategy—it's likely not worth the trouble.
* Add a test flag for forcing a rewrite of the store. Also revisit trying to get the SHA in the store directory name.
* Convert this single-node test to a 3-node test.

And probably more—but we're getting close to having a user-friendly TPCC microbenchmark!

---

>I thought --rewrite was no longer necessary?

Fixed.

>FYI there's also b.TempDir which gets auto-removed. (I think this didn't exist when testutils.TempDir was introduced).

Thanks for the pointer. I switched to this in a new commit

>+1 on fast feedback loops. I wasn't suggesting that this should become a heavyweight test. I find BenchmarkParallelSysbench quite useful, but I have also spent time on mutex contention, for which a single-threaded benchmark obviously makes no sense. Additionally, a concurrent benchmark can do a better job at highlighting the impact of garbage collection (since garbage is created at higher rates, and GC runs more frequently).
>
>We didn't add BenchmarkParallelSysbench at first either, though, so no pressure to do so now - but it would be good if there weren't obvious issues with doing so later that couldn't easily be solved now. I'm not aware of any.

👍 SGTM.

>I was actually worried about the opposite. Concurrency is admittedly a bit more heaviweight, but it also makes the benchmark more realistic. I was worried that if we added concurrency without also incrementing the warehouse count, we'd basically test the performance of incrementing a single key (since I think the order IDs are pulled from a per-warehouse counter). But since we're not adding concurrency now, we are fine with one warehouse.
And similar to sysbench, if we do add it, it can be under a new top-level test, and that one can be configured with more warehouses.

Good point. We'll keep this in mind when we add a parallel variant.

tbg commented:
Yes please merge and thank you for going through all of this!

On Fri, May 23, 2025, 23:53 Marcus Gartner ***@***.***> wrote:

> *mgartner* left a comment (cockroachdb/cockroach#147125)
> <https://protect.checkpoint.com/v2/r01/___https://github.com/cockroachdb/cockroach/pull/147125%23issuecomment-2905896300___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6ZzpkOTUzNTQ4NzljNzVkYjAxNWQ0Y2QyMjU5YTcxNmQzZjo3OmY5MTM6NDJmZTIzNTU4YzkzZGZiMzc3ZjA3YmYxNmQ4ZDAyM2M1ZTIwODg0ZGU4NDI0ZGM3MTY5MTFiMmE0NjkxYzgyZjpoOlQ6Tg>
>
> I think I've addressed all the minor feedback, and this takes us in the
> right general direction, so I'm going to merge this when I get a passing
> build. There's still more work to be done though! I'll have a follow-up
> PR(s) for the things discussed above:
>
>    - Removing the subprocess strategy—it's likely not worth the trouble.
>    - Add a test flag for forcing a rewrite of the store. Also revisit
>    trying to get the SHA in the store directory name.
>    - Convert this single-node test to a 3-node test.
>
> And probably more—but we're getting close to having a user-friendly TPCC
> microbenchmark!
> ------------------------------
>
> I thought --rewrite was no longer necessary?
>
> Fixed.
>
> FYI there's also b.TempDir which gets auto-removed. (I think this didn't
> exist when testutils.TempDir was introduced).
>
> Thanks for the pointer. I switched to this in a new commit
>
> +1 on fast feedback loops. I wasn't suggesting that this should become a
> heavyweight test. I find BenchmarkParallelSysbench quite useful, but I have
> also spent time on mutex contention, for which a single-threaded benchmark
> obviously makes no sense. Additionally, a concurrent benchmark can do a
> better job at highlighting the impact of garbage collection (since garbage
> is created at higher rates, and GC runs more frequently).
>
> We didn't add BenchmarkParallelSysbench at first either, though, so no
> pressure to do so now - but it would be good if there weren't obvious
> issues with doing so later that couldn't easily be solved now. I'm not
> aware of any.
>
> 👍 SGTM.
>
> I was actually worried about the opposite. Concurrency is admittedly a bit
> more heaviweight, but it also makes the benchmark more realistic. I was
> worried that if we added concurrency without also incrementing the
> warehouse count, we'd basically test the performance of incrementing a
> single key (since I think the order IDs are pulled from a per-warehouse
> counter). But since we're not adding concurrency now, we are fine with one
> warehouse.
> And similar to sysbench, if we do add it, it can be under a new top-level
> test, and that one can be configured with more warehouses.
>
> Good point. We'll keep this in mind when we add a parallel variant.
>
> —
> Reply to this email directly, view it on GitHub
> <https://protect.checkpoint.com/v2/r01/___https://github.com/cockroachdb/cockroach/pull/147125%23issuecomment-2905896300___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6ZzpkOTUzNTQ4NzljNzVkYjAxNWQ0Y2QyMjU5YTcxNmQzZjo3OjdjMTY6NjYyNDQxOWQ1OGY5ZjFmNTg1Mjk4YWNmNGNjZWNmOTliMjcyOWU5MjI1ZTY4MmZiYzIxNzJmNDNiMmUzMmU0YTpoOlQ6Tg>,
> or unsubscribe
> <https://protect.checkpoint.com/v2/r01/___https://github.com/notifications/unsubscribe-auth/ABGXPZCV7DIJ5FCAVPJQC5T276KEPAVCNFSM6AAAAAB5UNSIBGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDSMBVHA4TMMZQGA___.YzJ1OmNvY2tyb2FjaGxhYnMyOmM6ZzpkOTUzNTQ4NzljNzVkYjAxNWQ0Y2QyMjU5YTcxNmQzZjo3OmQ5NzU6MTNlZmNiNzY5ZDkzYjJkODk4Nzk2Y2JjNzM5NTAwNDIzYzc3YTY3YjkyNzllYWQ0NjYwNzA3MTg4ODU1MDMzNTpoOlQ6Tg>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>
